singledecoder.go

```diff
@@ -117,9 +118,12 @@ func (c *SingleDecoder) getProgram(ctx context.Context, vt reflect.Type, wID int
 	} else {
 		// We haven't seen the writer schema before, so try to fetch it.
 		wType, err = c.registry.SchemaForID(ctx, wID)
-		// TODO look at the SchemaForID error
-		// and return an error without caching it if it's temporary?
-		// See https://github.com/heetch/avro/issues/39
+		// do not cache the error when schema registry is unavailable
+		// we can't import avroregistry, to compare the error, so we're looking at the error message to see if the
+		// error is of type `UnavailableError` (avroregistry/errors.go)
+		if err != nil && strings.HasPrefix(err.Error(), "schema registry unavailability caused by") {
+			return nil, err
```

Comment:
Can it lead to an overload of the registry if we don't cache anything? The other open PR that addresses the same issue caches the error for 1 minute: https://github.com/heetch/avro/pull/127

Can we just merge that one, or maybe decrease the duration a bit if necessary?

Reply:
Merging mentioned PR would be nice.

1min might be a bit too much in our case indeed. How about we make that duration configurable (e.g w/ default and overridable through env var)?

Reply:
When the schema registr\y is unavailable the request doesn't reach the schema registry pod at all, so it's not stressed.

If i'm understanding this correctly, this should only be called once per pod per topic (unless there's a race condition where it's called for each partition on the same pod at the same time). In any case, the number of requests to the schema registry is very low.

Reply:
> When the schema registr\y is unavailable the request doesn't reach the schema registry pod at all, so it's not stressed.

If the registry is unavailable for 10 minutes, and in that timeframe one pod after another is started (e.g. autoscaling, node consolidation) and runs into the error, then the registry comes back up, won't all pods send their requests to the registry at the same time? Could that lead to issues?
If yes, the error caching could spread that initial load a bit if I'm not mistaken (assuming the pods with the errors didn't all start at the same time, but spread over those 10 minutes of registry downtime).

Reply:
Ideally we could check the exact issue
* unavailable registry issue: do not cache the error
* registry response issue (invalid schema, data store error, etc): cache the error for X minutes

But this current PR is good enough as a first step IMO

> won't all pods send their requests to the registry at the same time? Could that lead to issues?

This is already the current behavior when a deploy happens, the registry should accommodate this (as Adrian mentioned the number of requests per service should be low, basically 1 per topic consumed/produced in per instance)

Reply:
My PR explicitly only doesn't cache `avroregistry.UnavailableError`
Looking at the code, that can only happen if: 
* the schema registry isn't available at all (which should be fixed by an open PR in universe)
* or if the schema registry returns a 5xx response code

I believe that it would be a mistake to cache the error in either of those cases. If we want to limit the cache time for other types of errors, that would be beyond the scope of this PR.